core/transfer/src/main/java/org/eclipse/dataspaceconnector/transfer/core/transfer/AsyncTransferProcessManager.java

```diff
@@ -278,6 +291,26 @@ private void run() {
         }
     }
 
+    private int processCommandQueue() {
+        var batchSize = 5;
+        var commands = commandQueue.dequeue(batchSize);
+        commands.forEach(command -> {
+            var commandResult = commandRunner.runCommand(command);
+            if (commandResult.failed()) {
+                //re-queue if possible
+                if (command.canRetry()) {
+                    monitor.warning(format("Could not process command [%s], will retry. error: %s", command.getClass(), commandResult.getFailureMessages()));
+                    commandQueue.enqueue(command);
```

Comment:
I fear that the command cannot be re-enqueued as its future would be already `exceptionally complete` so, every other completion on it would not work. 
Personally I would completely remove the `CompletableFuture` from the command/commandrunner, also because it's not serializable so completely useless in a persisted queue scenario.

Reply:
Yup, that's why it has the `transient` modifier. We might still need a way of notifying the original creator of the command (i.e. the "client") about its completion. Classic callbacks have the same problem of non-serializability.